File: 3677.count-binary-palindromic-numbers.go
#
# @lc app=leetcode id=3677 lang=golang
#
# [3677] Count Binary Palindromic Numbers
#

# @lc code=start
func countBinaryPalindromes(n int64) int {
	if n == 0 {
		return 1
	}

	un := uint64(n)
	L := 0
	tmp := un
	for tmp > 0 {
		tmp >>= 1
		L++
	}

	ans := int64(1)
	for l := 1; l < L; l++ {
		h := (l + 1) / 2
		ans += int64(1 << (h - 1))
	}

	h := (L + 1) / 2
	m := L / 2
	leftMin := uint64(1) << uint(h - 1)
	for left := leftMin; left < (uint64(1) << uint(h)); left++ {
		num := left << uint(L - h)
		rev := uint64(0)
		for i := 0; i < m; i++ {
			if left & (uint64(1) << uint(h - 1 - i)) != 0 {
				rev |= uint64(1) << uint(i)
			}
		}
		num |= rev
		if num <= un {
			ans++
		} else {
			break
		}
	}

	return int(ans)
}
# @lc code=end

File: 3677.count-binary-palindromic-numbers.java
#
# @lc app=leetcode id=3677 lang=java
#
# [3677] Count Binary Palindromic Numbers
#

# @lc code=start
class Solution {
  public int countBinaryPalindromes(long n) {
    if (n == 0) return 1;
    String bin = Long.toBinaryString(n);
    int L = bin.length();
    long ans = 1;
    for (int len = 1; len < L; len++) {
      ans += 1L << ((len - 1) / 2);
    }
    int half = (L + 1) / 2;
    String prefix = bin.substring(0, half);
    long smaller = 0;
    for (int i = 0; i < half; i++) {
      if (prefix.charAt(i) == '1') {
        smaller += 1L << (half - i - 1);
      }
    }
    smaller -= 1L << (half - 1);
    long palin = 0;
    for (int i = 0; i < half; i++) {
      if (prefix.charAt(i) == '1') {
        palin |= (1L << i) | (1L << (L - 1 - i));
      }
    }
    long eq = (palin <= n) ? 1L : 0L;
    ans += smaller + eq;
    return (int) ans;
  }
}
# @lc code=end